### PR TITLE
Remove leftover load profile

### PR DESCRIPTION
**Pull request by unkcpz**
The `load_profile()` in `__init__.py` to load the profile in module scope was introduced by PR https://github.com/aiidalab/aiidalab-widgets-base/pull/12. It is not needed anymore since the computer and code widgets are updated and can update the dropdown dynamically. 

More problematically that this makes the profile loaded in the unit tests and overrides the test profile which makes it has to put the module load inside every test. There is also the problem that if the user want to test the widgets in a newly create python environment that doesn't have an AiiDA profile setup impossible. 

- [x] move modules import from test to file in front.

Before meeting this, we should make sure that apps which depend on AWB, load profiles in the notebooks. We checkbox the apps if they do:

- [x] [Aurora app](https://github.com/epfl-theos/aiidalab-aurora) (doesn't depend on AWB)
- [ ] [Molecules app](https://github.com/nanotech-empa/aiidalab-empa-molecules)
- [x] [LSMO app](https://github.com/lsmo-epfl/aiidalab-epfl-lsmo.git) (not migrated)
- [ ] [Martket place - use case 3](https://github.com/daniel-sintef/aiidalab-uc3)
- [x] [Graphene nanoribbons](https://github.com/nanotech-empa/aiidalab-empa-nanoribbons)
- [ ] [Quantum ESPRESSO app](https://github.com/aiidalab/aiidalab-qe)
- [x] [Scanning Probe Microscopy](https://github.com/nanotech-empa/aiidalab-empa-scanning-probe) (deprecated)
- [ ] [SSSP toolbox](https://github.com/aiidalab/aiidalab-sssp)
- [x] [On-Surface Chemistry](https://github.com/nanotech-empa/aiidalab-empa-surfaces)

### CODE REVIEW

**Comment by unkcpz** [comment]
well, there is one thing needs to change for the existing notebooks or apps the `load_profile` need to be explicitly called from app. Which also means the apps need to be adopted.
I think it is a more proper way to do it, as instructed in the aiida doc https://aiida.readthedocs.io/projects/aiida-core/en/latest/intro/installation.html?highlight=jupyter%20notebook#using-aiida-in-jupyter. 

pinning @yakutovicha @danielhollas for comments.

**Review by danielhollas — CHANGES_REQUESTED** [review]
I like this change. In particular, this will allow in the future to support multiple AiiDA profiles. 
However, as you note this will break most of the existing apps so we should not take this lightly. It's a pity that we did not do it before the 2.0 release. Should we wait with this for 3.0?

On a practical side, I would propose to add a code in `aiidalab_widgets_base/__init__.py` that would check whether the AiiDA profile is loaded and throws a `RuntimeError` if that is not the case. This error should at least make it very clear to the developers what needs to be done.

**Comment by unkcpz** [comment]
>  Should we wait with this for 3.0?

I don't think this will happen soon.  @yakutovicha we still have control of most of app, do you think it is doable to make this change? I'd argue this can be a feature that is worth adding with a minor release.

> I would propose to add a code in aiidalab_widgets_base/__init__.py that would check whether the AiiDA profile is loaded and throws a RuntimeError if that is not the case. 

I didn't see the point of doing this. This is to prevent the loaded profile being overrided by the `load_profile` of `aiidalab_widgets_base/__init__.py`? Changing/overriding the profile in a notebook is not a problem I guess?

**Comment by danielhollas** [comment]
> I didn't see the point of doing this. This is to prevent the loaded profile being overrided by the load_profile of aiidalab_widgets_base/__init__.py?

The point is only to provide a clear error message in case the developer forgets to call `load_profile()` before loading AWB. I do not know what happens now, perhaps the existing error that one gets is clear enough? 

The question I had not  considered is whether there are widgets which do not require loaded AiiDA profile? In which case we should not have this runtime check I guess.

**Comment by yakutovicha** [comment]
> I don't think this will happen soon. @yakutovicha we still have control of most of app, do you think it is doable to make this change? I'd argue this can be a feature that is worth adding with a minor release.

I don't think it is a big problem to do that. In most of our notebooks, we have `%aiida` magic which does that already.

At the moment, the only fear I have is that some of the modules of AWB that require AiiDA classes will fail to load. Maybe this is ok (in the end, for exactly that reason we launch `verdi shell` instead of just `ipython`).

**Comment by yakutovicha** [comment]
So I would probably go with this PR taking the suggestion of @danielhollas to through `RuntimeError` in case the profile is not loaded.

**Comment by unkcpz** [comment]
> load_profile() before loading AWB. I do not know what happens now, perhaps the existing error that one gets is clear enough?

As I did try in lastest two commits, if the check is put in the `__init__.py` it will run when the module is imported, and the exception will then raise if the profile is not loaded before `import aiidalab_widgets_base`. I think the current error is clear enough for AiiDA users. Therefore I rollback the change.

**Comment by yakutovicha** [comment]
@unkcpz, is there anything left here to be done?

**Comment by danielhollas** [comment]
Are we in a rush for this? This is a hard breaking change for essentially
all apps that use AWB.

Dne st 12. 7. 2023 9:13 uživatel Aliaksandr Yakutovich <
***@***.***> napsal:

**Comment by unkcpz** [comment]
> Are we in a rush for this? This is a hard breaking change for essentially

Not in a rush and yes it is a change that will break other apps. So I tag it to milestone `2.1.0`. But it needs to happen sometime, so the question to me is what is the blocker for not making this move? At least we still have all control of all apps and we are in the stage of migrating things, once all app migration in EMPA is finished or your app is released, there will be more resistance to pushing breaking changes, I guess?

> @unkcpz, is there anything left here to be done?

Yep, It is ready. I'll address your requests.

**Comment by unkcpz** [comment]
> We can, for example, list all the apps and tick-box them if they do.

This is a good idea, can you create a list of apps and tick boxes for EPMA apps? I'll check other apps.

**Comment by yakutovicha** [comment]
> > We can, for example, list all the apps and tick-box them if they do.
> 
> This is a good idea, can you create a list of apps and tick boxes for EPMA apps? I'll check other apps.

Sure, I will put it in the top comment.

**Comment by danielhollas** [comment]
> Not in a rush and yes it is a change that will break other apps. So I tag it to milestone 2.1.0. But it needs to happen sometime, so the question to me is what is the blocker for not making this move? At least we still have all control of all apps and we are in the stage of migrating things, once all app migration in EMPA is finished or your app is released, there will be more resistance to pushing breaking changes, I guess?

Note that I've already sent my app to some external collaborators so I'd really like things to be stable. At the very least, for breaking things like this I'd like us to get back into a habit of having a reasonable deprecation period.

The good news here is that we can first make all the existing apps ready for this change, by making sure they load the profile in the notebook (loading profile twice is okay). 

I would propose to first do:

1. In AWB `__init__`, check whether profile is loaded. If it is, do nothing.
2. If a profile is not loaded, load the default profile and emit a deprecation warning.

In this way, we can release this change in a fully backwards compatible manner (even in a patch release). At the same time it would already allow us to do all the things we wanted (e.g. load non-default profiles, easier unit testing).

WDYT @unkcpz @yakutovicha ?

**Comment by yakutovicha** [comment]
> I would propose to first do:
> 
> 1. In AWB `__init__`, check whether profile is loaded. If it is, do nothing.
> 2. If a profile is not loaded, load the default profile and emit a deprecation warning.

@unkcpz I assume you take care of that, right?

**Comment by unkcpz** [comment]
Yes, I'll take care of this one.

**Comment by danielhollas** [comment]
@unkcpz thanks! Since this is going to be backward compatible for now, I've moved it to 2.0.2 release.

File: aiidalab_widgets_base/__init__.py [hunk 1]
```diff
@@ -1,7 +1,39 @@
...
+    <pre style="background-color: #f0f0f0; padding: 10px; border: 1px solid #ccc; font-family: 'Courier New', monospace;">
 from aiida import load_profile
+load_profile();</pre>
+</div>
+"""
+
+# use the environment variable to control the automatic profile loading
+# so it can be disabled for testing, e.g. in the CI
+# use `TESTING=1 pytest` to disable the automatic profile loading
+TESTING = os.environ.get("TESTING", "False").lower() in ["true", "1"]
```
Comment (danielhollas) [inline aiidalab_widgets_base/__init__.py:19]: I am not a fan of this solution. Could we instead detect whether we run in the Jupyter notebook context (perhaps checking whether ipython is loaded or something like that?). That way we should be able to detect when we're unit testing and not load the profile.

**Review by danielhollas — CHANGES_REQUESTED** [review]
The warning message is very nice! Since it is rather intrusive, we should at least bump the minor version when we release this.

File: aiidalab_widgets_base/__init__.py (hunk 1 above)
Comment (unkcpz) [reply aiidalab_widgets_base/__init__.py]: I also adapt suggestions from @yakutovicha, the import is with module namespaces now.

Comment (danielhollas) [reply aiidalab_widgets_base/__init__.py]: Great! I like the solution with `is_running_in_jupyter`. :+1:

File: notebooks/aiida_datatypes_viewers.ipynb [hunk 2]
```diff
@@ -7,6 +7,17 @@
     "# Displaying AiiDA data types"
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from aiida import load_profile\n",
```
Comment (danielhollas) [inline notebooks/aiida_datatypes_viewers.ipynb:16]: Can you please revert this change, so that we test that the default profile is loaded in __init__ still?
Also, you could then extend the `test_aiida_datatypes_viewers` test in `tests_notebooks/test_notebooks.py` to check that the deprecation warning is displayed.

**Review by danielhollas — CHANGES_REQUESTED** [review]
@unkcpz thanks, this looks great now. I've tested this locally. I only have one request for testing the new behaviour.

This should be released in 2.1 version. So before merging, we should release 2.0.2. There are still some tasks in that milestone. I've taken care of one in #502.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -5,6 +5,12 @@ name: continuous-integration
 
 on: [push, pull_request]
 
+# https://docs.github.com/en/actions/using-jobs/using-concurrency
+concurrency:
```
[the remark this answers is not in the document]
Comment (unkcpz) [reply .github/workflows/ci.yml:9]: Is not needed here, but it is useful CI workflow configuration. Every time the new commit of a PR is submitted, it will stop the previously running test, this what we usually want and this save the time.

File: notebooks/aiida_datatypes_viewers.ipynb (hunk 2 above)
Comment (unkcpz) [reply notebooks/aiida_datatypes_viewers.ipynb:16]: Good point, I'll try to add a test.

**Review by yakutovicha — COMMENTED** [review]
LGTM,

The only thing I would consider is whether to propose to the developers an alternative solution using ipython magic:

```python
%load_ext aiida
%aiida
 ```

**Comment by unkcpz** [comment]
> The only thing I would consider is whether to propose to the developers an alternative solution using ipython magic:

```
%load_ext aiida
%aiida
```

Yes, we can, but I remember there are issues with this magic, although all are fixed (https://github.com/search?q=repo%3Aaiidateam%2Faiida-core+ipython+magic&type=issues) but it is less robust than the `load_profile` I'd say. So I would recommend using the load_profile.

**Review by yakutovicha — APPROVED** [review]
Great work, @unkcpz!

**Comment by unkcpz** [comment]
Thanks for the review! This will go to `v2.1` so anyone who makes the release should remember to cherry-pick commits when doing the release.
